File: starter_kits/Rust/src/hlt/position.rs
use hlt::direction::Direction;

#[derive(Copy, Clone, PartialEq, Eq)]
pub struct Position {
    pub x: i32,
    pub y: i32,
}

impl Position {
    pub fn directional_offset(&self, d: Direction) -> Position {
        let (dx, dy) = match d {
            Direction::North => (0, -1),
            Direction::South => (0, 1),
            Direction::East => (1, 0),
            Direction::West => (-1, 0),
            Direction::Still => (0, 0),
        };

        Position { x: self.x + dx, y: self.y + dy }
    }

    pub fn get_surrounding_cardinals(&self) -> Vec<Position> {
        vec! {
            self.directional_offset(Direction::North), self.directional_offset(Direction::South),
            self.directional_offset(Direction::East), self.directional_offset(Direction::West)
        }
    }
}

File: starter_kits/Kotlin/hlt/Position.kt
package hlt

class Position(val x: Int, val y: Int) {

    internal fun directionalOffset(d: Direction): Position {
        val dx: Int
        val dy: Int

        when (d) {
            Direction.NORTH -> {
                dx = 0
                dy = -1
            }
            Direction.SOUTH -> {
                dx = 0
                dy = 1
            }
            Direction.EAST -> {
                dx = 1
                dy = 0
            }
            Direction.WEST -> {
                dx = -1
                dy = 0
            }
            Direction.STILL -> {
                dx = 0
                dy = 0
            }
        }

        return Position(x + dx, y + dy)
    }
}
